### PR TITLE
Add ksmConfiguration to enable ksm in the nodes

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
With `ksmConfiguration` cluster admins are able to define a set of nodes in which ksm will be enabled.
The nodes will be filtered by label with the LabelSelector. If a node matches the labels and has the ksm available, the ksm in that node will be enabled. Alongside the enabling of the ksm, an annotation will be added to the node.
This annotation marks nodes where the ksm has been enabled by the virt-handler, allowing kv to disable it in case a subsequent change in the selector will cause a node to no longer be included.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Cluster admins can enable ksm in a set of nodes via kv configuration
```

/cc @xpivarc @acardace @vladikr @Barakmor1

### CODE REVIEW

**Comment by alicefr** [comment]
@fossedihelm I have a question. Why don't we use a label instead of an annotation? I wondering because if we use a label then we could use it to schedule the VM and filtering the nodes with ksm.

**Comment by fossedihelm** [comment]
> @fossedihelm I have a question. Why don't we use a label instead of an annotation? I wondering because if we use a label then we could use it to schedule the VM and filtering the nodes with ksm.

@alicefr A label for those nodes has been introduced in https://github.com/kubevirt/kubevirt/pull/9691 :)
This annotation is used to track which nodes has been modified by the virt-handler, so that the ksm will be disabled in case of changes in the selector terms. For example when the ksm is first enabled in all the nodes, but then a filter is introduced: in this case, in the nodes in which ksm was first enabled, it will be disabled. Hope this helps

**Comment by alicefr** [comment]
@fossedihelm sorry, I might missing an essential point but I fail to understand why we need 2 ways to mark the node. Isn't a single label sufficient? Or is it to mark which nodes have already been updated?

File: tests/infra_test.go
Comment (acardace) [inline tests/infra_test.go:1716]: Shouldn't we use Ginkgo labels/decorators instead of blindly skipping?

**Comment by acardace** [comment]
> @fossedihelm sorry, I might missing an essential point but I fail to understand why we need 2 ways to mark the node. Isn't a single label sufficient? Or is it to mark which nodes have already been updated?

Hi Alice, 
I'll try to answer, the annotation is an internal record to keep track of which nodes are currently managed by virt-handler (in regard to KSM), and the label is to be used by VMs for scheduling purposes.

The label itself could be present also on nodes where KSM was not enabled and thus managed by KubeVirt.

**Comment by alicefr** [comment]
@acardace  thanks :) it is clearer now

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go:2591]: IMHO `metav1.LabelSelector` allows to define more complex matching rules (i.e. the `In` operator will allow to do a sort of OR operation). Also, it is totally aligned with the k8s syntax

File: cmd/virt-handler/virt-handler.go
Comment (fossedihelm) [reply cmd/virt-handler/virt-handler.go:587]: Thanks, right, I will update it soon

File: tests/infra_test.go
Comment (fossedihelm) [reply tests/infra_test.go:1716]: I had thought about it but couldn't think of a way to activate that label. Now that you've pointed it out to me again, it's quite simple actually. Thanks

File: tests/infra_test.go
Comment (fossedihelm) [reply tests/infra_test.go]: Added

File: tests/infra_test.go
Comment (fossedihelm) [reply tests/infra_test.go:1717]: Done

File: cmd/virt-handler/virt-handler.go
Comment (fossedihelm) [reply cmd/virt-handler/virt-handler.go]: Done

File: cmd/virt-handler/virt-handler.go
Comment (fossedihelm) [reply cmd/virt-handler/virt-handler.go]: In order to avoid continuous logging, I moved the read only when we really want to enable or disable ksm

**Comment by fossedihelm** [comment]
@acardace I addressed the comments. PTAL. Thanks

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (acardace) [reply staging/src/kubevirt.io/api/core/v1/types.go:2591]: I Would still call it `NodeSelector` instead of just `Selector` it better conveys the purpose.

Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go]: thanks! If you agree, I opted for `NodeLabelSelector` :)

File: cmd/virt-handler/virt-handler.go
Comment (acardace) [inline cmd/virt-handler/virt-handler.go:587]: Here the log message should be something like `An error occurred while retrieving the node object` since it's a `Get()` and not a `List()`?

**Review by xpivarc — CHANGES_REQUESTED** [review]
There is a possibility that as this is written a lot of `get` calls could be done at the same time ( O(n) - n=#nodes). Could we rather re-use some of the already available routines? (node-labeler maybe)?

**Comment by fossedihelm** [comment]
> There is a possibility that as this is written a lot of `get` calls could be done at the same time ( O(n) - n=#nodes). Could we rather re-use some of the already available routines? (node-labeler maybe)?

I didn't realize that the node labeller immediately reacts to the config changes. I think it is better. Thank you!

File: automation/test.sh
Comment (xpivarc) [reply automation/test.sh:484]: By misconfiguration I mean a case where I forgot to enable KSM in the `project-infra`. With this, you would skip the tests but that would be wrong because I would rather get feedback, "You forgot about KSM!". Therefore all I am asking is rather not have this and move the responsibility of skip to elsewhere.

Comment (fossedihelm) [reply automation/test.sh:484]: Well, this is not related to enable ksm, but if it is available.
Are you suggesting to switch bask to `Skip()` inside the tests?

Comment (acardace) [reply automation/test.sh:484]: I think KSM is always compiled in the kernel for CentOS, so not having it would be pretty hard, that said here probably anything works, even just always adding the KSM ginkgo label without detecting anything.

Comment (xpivarc) [reply automation/test.sh:484]: >Are you suggesting to switch bask to Skip() inside the tests?

Not really. I am suggesting to not do any auto-detections. We should stick to what we do now, for example, add skip for lanes.
It is the same as you would be detecting if there is KVM or GPU... You never know if you did provision the right hardware or if the components that expose the resources just did not die. It is better to fail than a false skip.

Comment (fossedihelm) [reply automation/test.sh]: Got it, thanks!

**Comment by fossedihelm** [comment]
@xpivarc I moved the logic inside the node-labeller to save `get` api calls. PTAL
@acardace Since this is not what you approved anymore, I would ask you another review.
Thanks

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/node_labeller.go:443]: Is this needed? I believe `LabelSelectorAsSelector` will return a `selector` which will match nothing and therefore line 454 should take care of this.

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/node_labeller.go:441]: Please add unit tests.

**Comment by xpivarc** [comment]
/approve
The approach looks good to me

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (fossedihelm) [reply pkg/virt-handler/node-labeller/node_labeller.go:443]: Without this line 448 will generate a nil pointer dereference error.
```
selector, err := metav1.LabelSelectorAsSelector(ksmConfig.NodeLabelSelector)
```

**Comment by fossedihelm** [comment]
/hold
I notice that current implementation requires two cycles for enable ksm and label the node where ksm is enabled.
This can be improved

**Comment by fossedihelm** [comment]
@xpivarc @acardace  [Changes](https://github.com/kubevirt/kubevirt/compare/0a6c557507b0311ede92fe05bf0dbe58dbb8a99c..ad9330749cad1efe853fc5d45a9844f340215f2e)
- Moved the `loadKSM` function under the enable/disable ksm one (This allows to enable and add label in one shot).
- Switch using `n.KSM.SysfsFilePath` which was already there and allows us to create custom unit tests.
- Added unit tests
PTAL! Thanks

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (fossedihelm) [reply pkg/virt-handler/node-labeller/node_labeller.go:440]: Done

**Comment by fossedihelm** [comment]
@jean-edouard Can you have a look?

**Comment by fossedihelm** [comment]
Rebased

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (jean-edouard) [inline pkg/virt-handler/node-labeller/node_labeller.go:462]: Can't we run `loadKSM()` before `shouldEnableKSM()` and just check `n.KSM.Enabled` here?

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (jean-edouard) [inline staging/src/kubevirt.io/api/core/v1/types.go:2593]: Probably better to not do that, all other comments do start with a space

**Review by jean-edouard — COMMENTED** [review]
Looks good, just a couple comments

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go:2593]: Mmmmmh, this could be an outcome of a rebase. Thanks!

File: pkg/virt-handler/node-labeller/node_labeller.go
Comment (fossedihelm) [reply pkg/virt-handler/node-labeller/node_labeller.go]: Great catch! Thanks! But for this I also need to introduce `Available` field. I also arranged the function name. PTAL

**Review by jean-edouard — APPROVED** [review]
@acardace @xpivarc please ensure the latest version still makes sense
/hold

**Comment by acardace** [comment]
LGTM, @xpivarc I'll leave it to you to unhold.

**Comment by fossedihelm** [comment]
Thank you everyone for your review

**Pull request merged**
